kpet/utils.py

```diff
@@ -33,3 +37,106 @@ def get_template_content(filename, dbdir=None):
     template_path = get_template_path(filename, dbdir)
     with open(template_path) as file_handler:
         return file_handler.read()
+
+
+def print2(s):
+    sys.stderr.write(s + '\n')
+
+
+def smart_layout_insert_prefix(f_layout, obj):
+    if obj is None:
+        return None
+    prefix = os.path.dirname(os.path.realpath(f_layout))
+    return '%s/%s' % (prefix, str(obj))
+
+
+def smart_get_layout(f_layout):
+    text = ''
+    with open(f_layout, 'r') as f:
+        text = ''.join(f.readlines())
+
+    layout = json.loads(text, object_pairs_hook=collections.OrderedDict)
+    return layout
+
+
+def smart_get_patterns_bylayout(d_layout, f_layout):
+    l_patterns = []
+    for key in d_layout:
+        val = d_layout[key]
+        if val is None:
+            continue
+
+        s_file = smart_layout_insert_prefix(f_layout, val)
+        text = ''
+        with open(s_file, 'r') as f:
+            text = ''.join(f.readlines())
+
+        p = json.loads(text, object_pairs_hook=collections.OrderedDict)
+        for e in p['patterns']:
+            if e not in l_patterns:
+                l_patterns.append(e)
+
+    return l_patterns
+
+
+def smart_get_src_files(l_patch):
```

Comment:
Honestly, I don't understand what does this function at all :cry:. I'm going to try a guess... Passing a patch as parameter you get the files modified? I'd use unidiff rather than doing the parsing by hand.

Reply:
Yes. It does the same thing as `unidiff.PatchSet(data)`. I parse it by hand as (a) I don't know unidiff before as the package (python3-unidiff) is not installed by default; (b) it's quite easy to file a fake patch for testing the tool. e.g. A fake patch file looks like:
$ cat -n /tmp/foo.patch 
     1	diff --git a/fs/btrfs/ctree.h b/fs/btrfs/ctree.h
     2	...
     3	diff --git a/fs/btrfs/file.c b/fs/btrfs/file.c
     4	...
     5	diff --git a/fs/btrfs/ioctl.c b/fs/btrfs/ioctl.c
     6	...

Hence, currently I'd like to keep the code, will update to use `unidiff` in next version.